core/finality-grandpa/src/tests.rs

```diff
@@ -566,6 +573,89 @@ fn finalize_3_voters_no_observers() {
 		"Extra justification for block#1");
 }
 
+
+#[derive(Debug, Default)]
+pub struct TestPool {}
+
+impl PoolApi for TestPool {
+	type Api = TestPoolApi;
+	fn submit_one(
+		&self,
+		_at: &BlockId<Block>,
+		_xt: ExtrinsicFor<Self::Api>,
+	) -> result::Result<ExHash<Self::Api>, error::Error> {
+		unimplemented!()
+	}
+}
+
+pub struct TestPoolApi {
+	delay: Mutex<Option<std::sync::mpsc::Receiver<()>>>,
+}
+
+impl txpool::ChainApi for TestPoolApi {
```

Comment:
(e.g. if we didn't tightly couple to `PoolApi` we could have a much more useful `Report` hook for tests).